### ISSUE TITLE
Does this library supports binary_parameters ?

### ISSUE THREAD

**Issue by asafyish** [brianc/node-postgres#2266]
Hi,

I would like to use this package with pgBouncer in transaction mode but it seems that if I will try to use prepared statements it won't work, while I am ok with that (simply not use prepared statements), there are additional packages that use this library (https://github.com/graphile/worker for example), and they are using prepared statements, and therefor won't work correctly. If I understand correctly, supporting binary_parameters will allow this library, pgBouncer and prepared statements to play along.

Also, it seems like a nice performance improvement.

**Comment by boromisp** [comment]
The `binary_parameters` seems to be specific to one of the golang postgres drivers.
It looks like that the actual "binary" part of that function is not necessary for what you want, that is just an implementation detail of their single round-trip, unnamed prepared statement.

It might just work with node-postgres as-is, have you tried it?

**Comment by boromisp** [comment]
I've read [somewhere](https://github.com/lib/pq/issues/209#issuecomment-80840010) that `PQExecParams` from libpq does work with pgBouncer transaction mode.
The main difference between [that](https://github.com/postgres/postgres/blob/ca5e93f769e1bf5e90d080cfcc8c0368ef649a7c/src/interfaces/libpq/fe-exec.c#L1508) and the node-postgres implementation seems to be buffering: libpq buffers the all the messages belonging to the query, while node-postgres sends each message one by one. I don't think this should matter with TCP, but I haven't looked at the internals of pgBouncer.

**Comment by lastmjs** [comment]
Having something like this might allow `node-postgres` to work with the AWS RDS Proxy as well. Seems any kind of transaction pooling will not work with `node-postgres`'s prepared statements as they are, that's my hunch.

**Comment by lastmjs** [comment]
`node-postgres` definitely does not handle transaction pooling well with prepared statements. @boromisp I am very interested having single round-tripe unnamed prepared statements in `node-postgres`. Is this currently possible by using the native libpq bindings with `node-postgres`? Would I be able to configure the binary setting for libpq from node-postgres?

**Comment by boromisp** [comment]
The node-postgres [implementation](https://github.com/brianc/node-postgres/blob/95b5daadaade40ea343c0d3ad09ab230fa2ade4c/packages/pg/lib/query.js#L144) is single round-trip.
The issue discussed [here](https://github.com/lib/pq/issues/209) is that the golang driver had to wait for the responses for some of the intermediate messages, but node-postgres doesn't do that.

**Comment by lastmjs** [comment]
I'm going to attempt to get transaction pooling working with node-postgres and `pgBouncer`. Just to clarify, does node-postgres support unnamed prepared statements?

**Comment by boromisp** [comment]
Yes, the parameterized queries are usually unnamed prepared statements. The library creates a prepared statement, binds the parameters and exeutes it in a single step.

**Comment by NeoPhi** [comment]
@boromisp It looks like on https://github.com/brianc/node-postgres/blob/95b5daadaade40ea343c0d3ad09ab230fa2ade4c/packages/pg/lib/query.js#L54 if values were passed in it will do a prepare which to me seems to imply two round trips.

**Comment by boromisp** [comment]
If you follow the implementation of `prepare`:
https://github.com/brianc/node-postgres/blob/95b5daadaade40ea343c0d3ad09ab230fa2ade4c/packages/pg/lib/query.js#L180
you can see that it doesn't wait for a response between messages, but prepares the statement, binds the parameters and executes the query synchronously. I assume that this counts as a single round-trip. Whether or not it is compatible with transaction pooling I haven't tested.

**Comment by eugene-kim** [comment]
> I'm going to attempt to get transaction pooling working with node-postgres and `pgBouncer`. Just to clarify, does node-postgres support unnamed prepared statements?

@lastmjs any luck on this your efforts?

**Comment by lastmjs** [comment]
It did not work
